src/com/facebook/buck/maven/Resolver.java

```diff
@@ -98,75 +120,114 @@
+          dependencyBuilder.put(buildKey(dep.getArtifact()), dep);
+        }
+      } else {
+        Dependency dep = getDependencyFromString(artifact);
+        dependencyBuilder.put(buildKey(dep.getArtifact()), dep);
+      }
+    }
+    this.repos = repoBuilder.build();
+    this.specifiedDependencies = dependencyBuilder.build();
   }
 
-  public void resolve(String... mavenCoords) throws RepositoryException, IOException {
-    RepositorySystem repoSys = locator.getService(RepositorySystem.class);
-    RepositorySystemSession session = newSession(repoSys);
-
-    Set<String> specifiedArtifacts = new HashSet<>();
-    for (String coords : mavenCoords) {
-      specifiedArtifacts.add(coords);
-    }
+  public void resolve() throws RepositoryException, ExecutionException, InterruptedException,
+      IOException {
+    final RepositorySystem repoSys = locator.getService(RepositorySystem.class);
+    final RepositorySystemSession session = newSession(repoSys);
 
     ImmutableMap<String, Artifact> knownDeps = getRunTimeTransitiveDeps(
         repoSys,
         session,
-        mavenCoords);
+        specifiedDependencies.values());
 
     // We now have the complete set of dependencies. Build the graph of dependencies. We'd like
     // aether to do this for us, but it doesn't preserve the complete dependency information we need
     // to accurately construct build files.
-    MutableDirectedGraph<Artifact> graph = buildDependencyGraph(repoSys, session, knownDeps);
+    final MutableDirectedGraph<Artifact> graph = buildDependencyGraph(repoSys, session, knownDeps);
 
     // Now we have the graph, grab the sources and jars for each dependency, as well as the relevant
     // checksums (which are download by default. Yay!)
+    ImmutableSetMultimap<Path, Prebuilt> downloadedArtifacts =
+        downloadArtifacts(repoSys, session, graph);
 
-    Map<Path, SortedSet<Prebuilt>> buckFiles = new HashMap<>();
+    createBuckFiles(downloadedArtifacts);
+  }
 
-    for (Artifact artifact : graph.getNodes()) {
-      List<String> extraVisibility = specifiedArtifacts.contains(artifact.toString()) ?
-        this.visibility : ImmutableList.<String>of();
-      downloadArtifact(artifact, repoSys, session, buckFiles, graph, extraVisibility);
-    }
+  private ImmutableSetMultimap<Path, Prebuilt> downloadArtifacts(
+      final RepositorySystem repoSys,
+      final RepositorySystemSession session,
+      final MutableDirectedGraph<Artifact> graph) throws ExecutionException, InterruptedException {
+    ListeningExecutorService exec = MoreExecutors.listeningDecorator(
+        Executors.newFixedThreadPool(
+            Runtime.getRuntime().availableProcessors(),
+            new MostExecutors.NamedThreadFactory("artifact download")));
+
+    @SuppressWarnings("unchecked")
+    List<ListenableFuture<Map.Entry<Path, Prebuilt>>> results =
+        (List) exec.invokeAll(FluentIterable.from(graph.getNodes())
```

Comment:
ant is complaining that you're casting to a parametrised type without specifying the parameter.

`@SuppressWarnings("unchecked")` will let you cast to any parameter, but you still have to provide one.

Still, I don't understand how this is supposed to work: you're passing a list of callables instead of a list of futures?